### PR TITLE
resources, stdlib: Adding 'suite' category to gem5

### PR DESCRIPTION
**Pull request by Harshil2107**
Change-Id: I4a3231db4ab367c6e25bd97a67b92c2499330246

### CODE REVIEW

File: src/python/gem5/resources/suite.py
Comment (powerjg) [inline src/python/gem5/resources/suite.py:55]: This line is your problem

File: src/python/gem5/resources/suite.py
Comment (powerjg) [inline src/python/gem5/resources/suite.py]: I think the following would work:
```suggestion
        for workload in self.workload_obj_list:
            yield workload
```
I think that if you make `__iter__` a generator then you don't need `__next__`. I could be wrong, this is a bit confusing.

File: src/python/gem5/resources/suite.py
Comment (powerjg) [inline src/python/gem5/resources/suite.py]: Why does a suite require these parameters?  Resources don't require these parameters; they are optional.

Comment (Harshil2107) [reply src/python/gem5/resources/suite.py]: We did that because we need the id and version to use obtain resources. We were thinking that rather than making a `suite.py` we would add `SuiteResource` to the `resource.py` like other resources so that class constructor will not neet id and version.

File: src/python/gem5/resources/suite.py
Comment (powerjg) [reply src/python/gem5/resources/suite.py]: Don't forget to address this. It's still valid even though github says "outdated"

**Comment by Harshil2107** [comment]
The above tests will not run properly as workload changes are not merged yet, we are waiting for the workload changes to get merged first before pushing any other changes.

File: configs/test.py
Comment (powerjg) [inline configs/test.py]: Needs license block

File: configs/test.py
Comment (powerjg) [inline configs/test.py]: What is this print statement for?

File: configs/test.py
Comment (powerjg) [inline configs/test.py]: Can we choose different things here? I suggest "big" and "small"

File: src/python/gem5/resources/resource.py
Comment (powerjg) [inline src/python/gem5/resources/resource.py]: Remove debug statements.

File: src/python/gem5/resources/resource.py
Comment (powerjg) [inline src/python/gem5/resources/resource.py]: As previously suggested, I think you can do this:
```suggestion
    def __iter__(self) -> "SuiteResource":
        for workload in self.workload_obj_list:
            yield workload
```

File: src/python/gem5/resources/resource.py
Comment (powerjg) [inline src/python/gem5/resources/resource.py]:
```suggestion
        return {group for workload in self.workloads for group in workload["input_group"]}
```

File: configs/test.py
Comment (powerjg) [inline configs/test.py:1]: Maybe this file isn't meant to be included?

File: configs/test.py
Comment (Harshil2107) [reply configs/test.py]: This was just for checking if it gets the correct number of workloads, this is not a proper test.

File: src/python/gem5/resources/resource.py
Comment (Harshil2107) [reply src/python/gem5/resources/resource.py]: This change is done, we haven't pushed it yet, we checked and it works the same way.

File: configs/test.py
Comment (Harshil2107) [reply configs/test.py:1]: yeah, this file isn't meant to be included, we will rebase everything to a single commit before converting this to a normal pr from draft.

File: configs/test.py
Comment (Harshil2107) [reply configs/test.py]: This file will be removed, it was just for checking our initial implementation.

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py:659]: This is missing a doc-string. The dictionary explaining the `workloads` is necessary.

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py]: Maybe all these should be private?

```suggestion
        self._resource_version = resource_version
        self._workloads = workloads
        self._workload_obj_list = []
        self._index = 0
```

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py]: Is there a reason we don't use "obtain_resource" instead? This isn't getting all the information about a resource we may need.

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py]: Doesn't this return a Workload?

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py:703]: Doc-string for all public functions at least.

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py]: How does this work if you haven't obtained the resources?

**Review by BobbyRBruce — CHANGES_REQUESTED** [review]

File: src/python/gem5/resources/resource.py
Comment (Harshil2107) [reply src/python/gem5/resources/resource.py]: Yes, we should be using `obtain_resource`. For someone who wants to create their own `SuiteResource` how do we handle the call to `obtain_resource`? I can think of 2 possible solutions:

- We ask users to create their own `resources.json` and pass it in as a client in ontain_resources, We can also pass the `clients` parameter to the `SuiteResource` constructor so that users can define which clients they want.

- We can make getters and setters for the `workload_obj_list` to allow users to directly edit the workload list. This needs some sort of way to not call `obtain_resources` when the user is using their resources. 

I am leaning towards the first one.

File: src/python/gem5/resources/resource.py
Comment (Harshil2107) [reply src/python/gem5/resources/resource.py]: The return type should be a `Generator[YieldType, SendType, ReturnType]` (documentation: https://docs.python.org/3/library/typing.html#typing.Generator). In our case, it would be `Generator["WorkloadResource", None, None]`.We will change it to the correct type.

**Comment by kunpai** [comment]
TODO: Rebase all commits into one commit before merging

File: src/python/gem5/resources/resource.py
Comment (powerjg) [inline src/python/gem5/resources/resource.py]: How about "get_input_groups" since it returns a `Set` not a `List`

File: src/python/gem5/resources/resource.py
Comment (powerjg) [inline src/python/gem5/resources/resource.py]:
```suggestion
        for info, workload in zip(self._workloads, self._workload_obj_list):
```

**Review by powerjg — COMMENTED** [review]
A couple more small things :)

File: src/python/gem5/resources/resource.py
Comment (kunpai) [inline src/python/gem5/resources/resource.py]: 'Optional parameter'

File: tests/pyunit/stdlib/resources/pyunit_suite_checks.py
Comment (kunpai) [inline tests/pyunit/stdlib/resources/pyunit_suite_checks.py]: "Optional parameter"

File: configs/example/gem5_library/riscvmatched-microbenchmark-suite.py
Comment (BobbyRBruce) [inline configs/example/gem5_library/riscvmatched-microbenchmark-suite.py]: I thought we had this setup so a suite was an iterator of Workloads. Shouldn't this be `for resource in microbenchmarks:`?

File: configs/example/gem5_library/riscvmatched-microbenchmark-suite.py
Comment (BobbyRBruce) [inline configs/example/gem5_library/riscvmatched-microbenchmark-suite.py:48]: This suite doesn't exist yet, right?

File: configs/example/gem5_library/riscvmatched-microbenchmark-suite.py
Comment (BobbyRBruce) [inline configs/example/gem5_library/riscvmatched-microbenchmark-suite.py:1]: What is the purpose of this example. I'd suggest adding a doc-string explaining.

Also, it should have a test which regularly runs to ensure it is still a functioning example, such as that in https://github.com/gem5/gem5/tree/48a40cf2f5182a82de360b7efa497d82e06b1631/tests/gem5/gem5_library_example_tests

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py]: `local_path` here confuses me. Surely suites don't have one? It's only resources that do.

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py]: There's a leading underscore here before `workload_obj_list`

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py]: Is there a reason this is optional? What does `None` mean which isn't the same as an empty list?

Ditto for `workloads`.

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py]: Why do we have both `workloads` and `workload_obj_list`. Surely `workload_obj_list` encapsulates all the data in `workloads`? I think we should just have `workload_obj_list` and rename it as `workloads`. I don't know why the current design is like this but, as a user, how am i supposed to know which of these parameters to use to specify the workloads in a suite?

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py]: Technically this doesn't return, it `yields`. So I think the syntax is `:yields: ...`

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py]: Related to what I discuss above, this optional parameter is weird. I'd add the resource id and version to the `WorkloadResource` and work from there.

Also, the Input group is in the suite, right? Not the workload itself, so we can still do this lookup.

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py:659]: Ok, having read through the rest of this code, I think one of the reasons you have this `workload` argument is you need a way of mapping `input_groups` to Workloads. Why not do this:

```py
workloads: Dict[ResourceWorkload, Set[str]]
```

Just a simple dictionary. The key being the Workload, and it's value being a set of input_groups. Then everything can be in one argument.

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py]: This isn't necessary, the class already implements `__iter__` which does exactly this.

Also, `get_all_workloads` sounds like it should return a list or set of workloads, not be a generator (in this case I'd expect it be called something like `next_workload`.)

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py:703]: I think there should be an exception thrown here if an input_group is specified which doesn't exist for this workload. Quite simple logic:

```py
if input_group not in get_input_groups():
    raise Exception("...")
```

I'd have the exception say that the value is invalid and then list the valid values (i.e., `get_input_groups()`).

**Review by BobbyRBruce — CHANGES_REQUESTED** [review]
I didn't go over the tests very much as i had some feedback on the broader APIs and funcitonalities

File: configs/example/gem5_library/riscvmatched-microbenchmark-suite.py
Comment (kunpai) [reply configs/example/gem5_library/riscvmatched-microbenchmark-suite.py:48]: Yes it does not, yet

File: configs/example/gem5_library/riscvmatched-microbenchmark-suite.py
Comment (Harshil2107) [reply configs/example/gem5_library/riscvmatched-microbenchmark-suite.py]: Yes, that also works but the above function returns a tuple with the workload id, version, input groups and the WorkloadResource object.

File: src/python/gem5/resources/resource.py
Comment (Harshil2107) [reply src/python/gem5/resources/resource.py]: As you specified below, the `workloads` field is mainly to keep track of the id, version, and input_group and link it to the workload object.

File: src/python/gem5/resources/resource.py
Comment (Harshil2107) [reply src/python/gem5/resources/resource.py]: Should we return a list of tuples or just change the function name?

Comment (powerjg) [reply src/python/gem5/resources/resource.py]: I think this function serves a different purpose than `__iter__` since it also returns the id, version, and inputs. 

I also think we're doing a bit of bikeshedding here.

But, to argue about the color of the shed... One option would be to make this the implementation of `__iter__` and to have this function just return the list of workloads.

IMO, the decision here doesn't really matter as long as we pick something that is consistent and we document it.

**Comment by kunpai** [comment]
> TODO: Rebase all commits into one commit before merging

Has been done :)

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py]: The `workloads` param doc-string is no longer needed here.

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py]: A better solution `workloads: Dict["WorkloadResource", Set[str]] = {}`.

Improvements:

1. Rename `workload_obj_list` to `workloads`.
2. Removes the confusing "List of tuples".
3. Allows for quick association of workload to group_label.

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [reply src/python/gem5/resources/resource.py]: Just get rid of this function. It's weird. If we need the id you can just do `workload.get_id() for workload in suite`. I really don't see any advantage of this function. The workloads which return via the iterator contain all the information you'd ever require.

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py]: Remove this newline.

**Review by BobbyRBruce — CHANGES_REQUESTED** [review]

**Comment by Harshil2107** [comment]
I made an error while trying to sync the new develop branch with my working branch, I will try to fix it in the morning.

File: configs/example/gem5_library/riscvmatched-microbenchmark-suite.py
Comment (BobbyRBruce) [reply configs/example/gem5_library/riscvmatched-microbenchmark-suite.py]: This will fail now. You need to change it.

File: configs/example/gem5_library/riscvmatched-microbenchmark-suite.py
Comment (BobbyRBruce) [reply configs/example/gem5_library/riscvmatched-microbenchmark-suite.py:1]: You added the doc-string here but i think it should be added as a test, as I suggested.

File: src/python/gem5/resources/resource.py
Comment (BobbyRBruce) [inline src/python/gem5/resources/resource.py]: There's a weird added line here. Remove it.

File: configs/example/gem5_library/riscvmatched-microbenchmark-suite.py
Comment (BobbyRBruce) [reply configs/example/gem5_library/riscvmatched-microbenchmark-suite.py:48]: Can we add it to make sure this script works as intended?

File: tests/pyunit/stdlib/resources/pyunit_suite_checks.py
Comment (BobbyRBruce) [inline tests/pyunit/stdlib/resources/pyunit_suite_checks.py]: This will fail, no? It's called `Workload` now.

**Review by BobbyRBruce — CHANGES_REQUESTED** [review]

**Review by BobbyRBruce — APPROVED** [review]

**Pull request merged**
